clean/platforms/nextrequest.py

```diff
@@ -0,0 +1,363 @@
+            if total_documents > profile["doc_limit"]:
+                message = f"Request found with {total_documents:,} documents, exceeding limits. "
+                message += f"This is probably a bad URL that can't be properly scraped: {page_url}. "
+                message += "Dropping record."
+                logger.warning(message)
+                returned_json = {}
+                file_needs_write = False
+                return (filename, returned_json, file_needs_write)
+            if max_pages > 1:
+                logger.debug(f"Need to download {max_pages - 1:,} more JSON files.")
+                for page_number in range(2, max_pages + 1):
+                    page_url = f"{json_url}{page_number}"
+                    if page_number >= 200:
+                        message = "NextRequest at least on some sites appears to have a hard limit of "
+                        message += f"199 pages. Not trying to scrape {page_url}."
+                        logger.warning(message)
+                    else:
+                        r = utils.get_url(page_url)
+                        if not r.ok:
+                            logger.error(
+                                f"Problem downloading {page_url}: {r.status_code}"
+                            )
+                            returned_json = {}
+                            file_needs_write = False
+                        else:
+                            additional_json = r.json()
+                            if "documents" not in additional_json:
+                                logger.error(
+                                    f"Missing 'documents' section from {page_url}"
+                                )
+                                returned_json = {}
+                                file_needs_write = False
+                            else:
+                                for i, _entry in enumerate(
+                                    additional_json["documents"]
+                                ):
+                                    additional_json["documents"][i][
+                                        "bln_page_url"
+                                    ] = page_url
+                                    additional_json["documents"][i][
+                                        "bln_total_documents"
+                                    ] = total_documents
+                                returned_json["documents"].extend(
+                                    additional_json["documents"]
+                                )
+                        sleep(throttle)
+            documents_found = len(returned_json["documents"])
+            if documents_found != total_documents:
+                message = f"Expected {total_documents:,} documents "
+                message += f"but got {documents_found:,} instead for "
+                message += f"{start_url}."
+                logger.warning(message)
+
+    return (filename, returned_json, file_needs_write)
```

Comment:
Nested conditionals can be rather gnarly and buggy to maintain over time, and they are often the first platform code to be rewritten.

I let gen-AI take a swing at an initial refactor for the purposes of pseudo-code. In other words, pseudo-code is an untested suggestion, not necessarily for adoption, to help rethink through the logic flow differently

```py
# UNTESTED GEN-AI PSEUDO-CODE FOR ILLUSTRATION
def fetch_nextrequest(
    base_directory: Path, start_url: str, force: bool = False, throttle: int = 2
) -> Tuple[Path, Union[None, Dict], bool]:
    """Given a link to a NextRequest documents folder, return a proposed filename and the JSON contents."""
    profile = fingerprint_nextrequest(start_url)
    folder_id = profile["folder_id"]
    json_url = profile["json_url"]

    local_cache = Cache(path=None)
    filename = base_directory / f"{folder_id}.json"
    if not force and local_cache.exists(filename):
        logger.debug(f"File found in cache: {filename}")
        return filename, None, False

    returned_json = download_json_data(json_url, profile, throttle)
    if not returned_json:
        return filename, {}, False

    file_needs_write = True
    return filename, returned_json, file_needs_write


def download_json_data(
    json_url: str, profile: Dict, throttle: int
) -> Union[None, Dict]:
    """Download JSON data from the given URL with pagination handling."""
    page_number = 1
    page_url = f"{json_url}{page_number}"
    r = utils.get_url(page_url)
    if not r.ok:
        logger.error(f"Problem downloading {page_url}: {r.status_code}")
        return None

    returned_json = r.json()
    total_documents = returned_json[profile["tally_field"]]
    add_metadata_to_documents(returned_json["documents"], page_url, total_documents)

    if total_documents > profile["doc_limit"]:
        logger.warning(
            f"Request found with {total_documents:,} documents, exceeding limits. Dropping record."
        )
        return {}

    max_pages = find_max_pages(total_documents, profile["page_size"])
    if max_pages > 1:
        logger.debug(f"Need to download {max_pages - 1:,} more JSON files.")
        for page_number in range(2, max_pages + 1):
            if page_number >= MAX_PAGES_LIMIT:
                logger.warning(
                    f"Not trying to scrape {json_url}{page_number} due to page limit."
                )
                break
            page_url = f"{json_url}{page_number}"
            r = utils.get_url(page_url)
            if not r.ok:
                logger.error(f"Problem downloading {page_url}: {r.status_code}")
                return None
            additional_json = r.json()
            if "documents" not in additional_json:
                logger.error(f"Missing 'documents' section from {page_url}")
                return None
            add_metadata_to_documents(
                additional_json["documents"], page_url, total_documents
            )
            returned_json["documents"].extend(additional_json["documents"])
            sleep(throttle)

    documents_found = len(returned_json["documents"])
    if documents_found != total_documents:
        logger.warning(
            f"Expected {total_documents:,} documents but got {documents_found:,} instead for {json_url}."
        )
    return returned_json


def add_metadata_to_documents(
    documents: List[Dict], page_url: str, total_documents: int
) -> None:
    """Add metadata to each document."""
    for doc in documents:
        doc["bln_page_url"] = page_url
        doc["bln_total_documents"] = total_documents
```